rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,571 @@
+
+These actions are always taken by a logged-in user (generally an inspector).
+
+There are additional automated processes which keep the schedule up-to-date:
+- importing date/schedule mappings
+- importing runs
+- importing scheduled trips
+
+The stream of events proposed by this RFC serves two purposes:
+- primarily, improve the rider-facing predictions produced by RTR, especially for departure times at a terminals, which currently do not have departure predictions at all, and
+- secondarily, other internal operations needs such as Lean or LAMP
+
+The goal of this RFC is to provide an interface which represents all data currently (2023H1) in Glides. While some additional flexibility is included, features outside the current scope of Glides are also outside the scope of these events and this RFC. 
+
+## CloudEvents
+All events will be in the [CloudEvents v1.0.2](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md) format (or later), using the JSON encoding. The event types will be under the `com.mbta.ctd.glides` event namespace, and will be in the past-tense (i.e. `trip-added` rather than `add-trip`).
+
+## Kinesis
+Events will be written to a Kinesis stream. Each Glides environment (`dev`, `dev-green`, and `prod`) will have a separate stream for their events, named `ctd-glides-<environment>`. 
+
+The partition key will be a hash of the station at which the inspector is working and the inspector's identity, which ensures that multiple events from a single inspector are ordered correctly if the messages are distributed across multiple shards. 
+As the stream will include potential PII (operator badge information) it should be configured as encrypted-at-rest. 
+
+Multiple events can be included in a single Kinesis record, by wrapping them in a JSON array. This is an optimization for improving write speeds: further steps in the Kinesis pipeline may break up or rearrange these arrays. If multiple events are part of the same user action, the [Batched][#Batched] event should be used to combine them.
+
+## Value types
+
+### Author
+Each message will include an `author` key, indicating who performed the given intervention.
+
+- `username` (string): the username of the logged-in user
+- `badgeNumber` (string, optional): the badge number of the logged-in user
+- `location` (Location, optional): the location the logged-in user is managing
+```json
+{
+  "username": "example",
+  "badgeNumber": "123456",
+  "location": {"gtfsId": "place-lake"}
+}
+
+or
+
+{
+  "username": "pswartz",
+  "location": {"gtfsId": "place-matt"}
+}
+```
+
+### Car
+Cars (individual parts of a train consist) will be described by their 4 digit number.
+```json
+{
+  "carNumber": "3800"
+}
+```
+It is represented as an object to provide future extensibility if needed.
+
+### Location
+One of:
+- `{"gtfsId": "<GTFS stop ID>"}`, where the GTFS `location_type` is 0 (stop) or 1 (station)
```

Comment:
This was added in response to this thread with @skyqrose: https://github.com/mbta/technology-docs/pull/18#discussion_r1083578333

Reply:
Sky did not say that the ID should be either a station or a platform. What they said is that the spec should clarify that the ID can only be a station. I'm not sure how you came to the conclusion that the ID could also be a platform. (The term "stop" is ambiguous here, since it can refer to either any location or to specifically a platform.)

Reply:
I interpreted their comment the other way, but I don't think that's particularly important. Is there a reason to restrict the field to only stations?

Reply:
I think the additional flexibility at this stage is worth it. 
- you'll need to create new versions for nearly all the events if you do want to support platforms in this field in the future (since adding a value here is not backwards compatible) 
- consumers can easily go from a platform to the parent station by consulting GTFS or the V3 API

Reply:
Yeah, I don't think this should allow child stops. We don't use child stops right now, and I can't think of any future inspector-entered data where we'd want them to have to think about selecting a specific platform instead of getting that from realtime data. And having only parent stations will make consumption easier.

If in the future we decide that we want to publish child stops, we can consider that a breaking change.